pkg/objstore/objstore.go

```diff
@@ -172,29 +232,45 @@ func NopCloserWithSize(r io.Reader) io.ReadCloser {
 
 // UploadDir uploads all files in srcdir to the bucket with into a top-level directory
 // named dstdir. It is a caller responsibility to clean partial upload in case of failure.
-func UploadDir(ctx context.Context, logger log.Logger, bkt Bucket, srcdir, dstdir string) error {
+func UploadDir(ctx context.Context, logger log.Logger, bkt Bucket, srcdir, dstdir string, options ...UploadDirOption) error {
 	df, err := os.Stat(srcdir)
+	opts := applyUploadDirOptions(options...)
+	g, ctx := errgroup.WithContext(ctx)
+	guard := make(chan struct{}, opts.concurrency)
```

Comment:
`errgroup.SetLimit` exactly for limiting the number of active goroutines. I think we should use it rather than implement our own mechanism.
Were there any reasons why you preferred it this way?

Reply:
good catch! i Will change

Reply:
Good catch! i Will change